### PR TITLE
core(inputs): refactor form-elements gatherer

### PR DESCRIPTION
**Pull request by connorjclark**
Refactors + renames the form-elements gatherer.

Instead of structuring the input elements inside form objects, place each input/label/form object into top-level properties and use indices to connect an input element with a form/labels.

It is still possible to query the artifact for "formless inputs" or "inputless labels" (if we ever wish to).

### CODE REVIEW

File: lighthouse-cli/test/smokehouse/core-tests.js [hunk 1]
```diff
@@ -72,8 +72,7 @@ const smokeTests = [
   errorsExpiredSsl,
   errorsIframeExpiredSsl,
   errorsInfiniteLoop,
-  // TODO: restore when --enable-features=AutofillShowTypePredictions is not needed.
-  // formsAutoComplete,
+  formsAutoComplete,
```
Comment (connorjclark) [inline lighthouse-cli/test/smokehouse/core-tests.js:75]: I've enabled the smoke test, but commented out the stuff that rely on the predictions being available.

File: lighthouse-core/audits/autocomplete.js [hunk 2]
```diff
@@ -224,47 +224,44 @@ class AutocompleteAudit extends Audit {
    * @return {LH.Audit.Product}
    */
   static audit(artifacts) {
-    const forms = artifacts.FormElements;
     const failingFormsData = [];
     const warnings = [];
     let foundPrediction = false;
-    for (const form of forms) {
```
Comment (connorjclark) [inline lighthouse-core/audits/autocomplete.js:231]: just whitespace changes, I didn't modify the logic.

File: lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js [hunk 3]
```diff
@@ -22,575 +21,570 @@ const config = {
...
+          labelIndicies: [
+            2,
+          ],
+          id: 'CCExpiresMonth1',
+          name: 'CCExpiresMonth1',
+          type: 'select-one',
+          autocomplete: {
+            property: 'cc-exp-month',
+            attribute: 'cc-exp-month',
+            prediction: null,
```
Comment (paulirish) [inline lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js:70]: all predictions are  now null. lets add a comment.. somewhere in this file... to clarify that they'd be populated if the flag were used.

**Review by paulirish — APPROVED** [review]
nice refactor. feels better

File: lighthouse-core/gather/gatherers/inputs.js [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+    // form= attribute is associated with a <form> element's id), this will be set.
+    const parentFormEl = inputEl.form;
+
+    inputArtifacts.push({
+      parentFormIndex: parentFormEl ?
+        [...formElToArtifact.keys()].indexOf(parentFormEl) :
+        undefined,
+      labelIndicies: [...inputEl.labels || []].map((labelEl) => {
+        return [...labelElToArtifact.keys()].indexOf(labelEl);
+      }),
```
Comment (adamraine) [inline lighthouse-core/gather/gatherers/inputs.js:65]: Nit: For readability, can you compute these on their own lines

File: types/artifacts.d.ts [hunk 5]
```diff
@@ -787,22 +787,22 @@ declare module Artifacts {
...
-    attributes?: {
-      id: string;
-      name: string;
-      autocomplete: string;
-    };
+  interface FormElement {
+    id: string;
+    name: string;
+    autocomplete: string;
     node: NodeDetails | null;
```
Comment (adamraine) [inline types/artifacts.d.ts:794]: Is there a situation where this is `null`?

File: lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js (hunk 3 above)
Comment (connorjclark) [reply lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js:70]: There's a comment below that mentions the flag

File: lighthouse-core/gather/gatherers/inputs.js [hunk 6]
```diff
@@ -0,0 +1,111 @@
+/**
+ * @license Copyright 2020 The Lighthouse Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (connorjclark) [reply lighthouse-core/gather/gatherers/inputs.js:2]: it's a file rename

[the remark this answers is not in the document]
Comment (adamraine) [reply lighthouse-core/gather/gatherers/inputs.js:2]: There are enough changes to consider this a new file imo, but I'll leave it up to you

**Review by adamraine — APPROVED** [review]

File: lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js (hunk 3 above)
Comment (paulirish) [reply lighthouse-cli/test/smokehouse/test-definitions/forms-autocomplete.js:70]: i saw that. but it's for the section you commented out. :)
i don't think the comment on line 467 is gonna help give context to whats happening on line 70.. 
so i was suggesting an additional comment.

**Pull request merged**